sweep/sweeper.go

```diff
@@ -414,57 +357,83 @@ func (s *UtxoSweeper) collector() {
 			// be started when new inputs arrive.
 			s.timer = nil
 
-			// Examine pending inputs and try to construct one or
-			// more sweep txes. Publish a tx even if it is not full
-			// yet (input count less than maximum allowed inputs per
-			// tx).
-			_, err := s.sweepAll(false)
+			// Retrieve fee estimate for input filtering and final
+			// tx fee calculation.
+			satPerKW, err := s.cfg.Estimator.EstimateFeePerKW(
+				s.cfg.SweepTxConfTarget)
+			if err != nil {
+				return err
+			}
+
+			// Examine pending inputs and try to construct lists of
+			// inputs.
+			inputLists, err := s.getInputLists(bestHeight, satPerKW)
 			if err != nil {
 				log.Errorf("Cannot sweep on timer event: %v",
 					err)
 			}
 
+			// Sweep selected inputs.
+			for _, inputs := range inputLists {
+				err := s.sweep(inputs, satPerKW, bestHeight)
+				if err != nil {
+					return err
+				}
+			}
+
+		case epoch := <-blockEpochs.Epochs:
+			bestHeight = epoch.Height
+			s.scheduleSweep(bestHeight)
 		case <-s.quit:
-			return
+			return nil
 		}
-
-		// TODO: Handle the case where a tx is not confirmed within
-		// specified time and fee needs to be bumped.
-
-		// TODO: Periodically refresh fee rate and see if it unlocks any
-		// inputs that then have a positive yield.
 	}
 }
 
-// scheduleSweep tries to sweep inputs, but only goes through if the generated
-// tx has the maximum number of inputs. If inputs remain, the batch timer is
-// started to create an opportunity for more inputs to be added.
-func (s *UtxoSweeper) scheduleSweep() {
-	// Try to sweep inputs, but only if full txes can be generated.
-	inputsRemaining, err := s.sweepAll(true)
+// scheduleSweep starts the sweep timer to create an opportunity for more inputs
+// to be added.
+func (s *UtxoSweeper) scheduleSweep(currentHeight int32) error {
+	// Retrieve fee estimate for input filtering and final
+	// tx fee calculation.
+	satPerKW, err := s.cfg.Estimator.EstimateFeePerKW(
+		s.cfg.SweepTxConfTarget)
 	if err != nil {
-		log.Errorf("Cannot sweep during scheduling: %v", err)
-		return
+		return err
 	}
 
-	if !inputsRemaining {
-		// Nothing sweepable anymore, disable timer. This prevents two
-		// sweeps from happening in fast succession. The first sweep
-		// triggered by the maximum allowed number of inputs and the
-		// second because of the timer.
-		log.Debugf("No inputs remaining, disabling timer")
-		s.timer = nil
-		return
+	// Examine pending inputs and try to construct lists of
+	// inputs.
+	inputLists, err := s.getInputLists(currentHeight, satPerKW)
+	if err != nil {
+		log.Errorf("Cannot sweep on timer event: %v",
+			err)
 	}
 
-	if s.timer == nil {
-		// Start sweep timer to create opportunity for more inputs to be
-		// added before a tx is constructed.
-		log.Debugf("Starting timer")
-		s.timer = s.cfg.NewBatchTimer()
-	} else {
+	// If there are inputLists, there is something to sweep.
+	sweepableInputs := len(inputLists) > 0
```

Comment:
Only slightly related to this new approach, but why remove the ability to hold off on a sweep unless a full transaction can be generated? A less restrictive from of that also seems useful, so in this form, we can provide a minimum number of inputs to keep our batch ratio high, and only force a sweep if we didn't get any new inputs before the next block arrived.